File: sol/1129.py
import heapq as hq
from collections import defaultdict

class Solution:
    def shortestAlternatingPaths(self, n: int, redEdges: List[List[int]], blueEdges: List[List[int]]) -> List[int]:
        rEdgeList = defaultdict(list)
        bEdgeList = defaultdict(list)
        for u, v in redEdges:
            rEdgeList[u].append(v)
        for u, v in blueEdges:
            bEdgeList[u].append(v)
        rDist = [-1] * n
        bDist = [-1] * n
        h = [(0, 0, False), (0, 0, True)]
        hq.heapify(h)
        while len(h) != 0:
            d, node, isRed = hq.heappop(h)
            if (rDist if isRed else bDist)[node] != -1:
                continue
            (rDist if isRed else bDist)[node] = d
            if isRed:
                for v in bEdgeList[node]:
                    if bDist[v] == -1:
                        hq.heappush(h, (d + 1, v, False))
            else:
                for v in rEdgeList[node]:
                    if rDist[v] == -1:
                        hq.heappush(h, (d + 1, v, True))
        res = [-1] * n
        for i in range(n):
            if rDist[i] == -1:
                res[i] = bDist[i]
            elif bDist[i] == -1:
                res[i] = rDist[i]
            else:
                res[i] = min(rDist[i], bDist[i])
        return res

File: sol/1129/1129.cpp
#include <unordered_map>
#include <queue>

struct Solution {
    vector<int> shortestAlternatingPaths(int n, vector<vector<int>>& redEdges, vector<vector<int>>& blueEdges) {
        unordered_map<int, vector<int>> rEdgeList, bEdgeList;
        for (auto &e : redEdges){
            rEdgeList[e[0]].push_back(e[1]);
        }
        for (auto &e : blueEdges){
            bEdgeList[e[0]].push_back(e[1]);
        }
        queue<int> q;
        queue<bool> stq;
        q.push(0); q.push(0); stq.push(false);
        q.push(0); q.push(0); stq.push(true);
        vector<int> rDist(n, -1), bDist(n, -1);
        while (!q.empty()){
            int d = q.front(); q.pop();
            int node = q.front(); q.pop();
            bool isRed = stq.front(); stq.pop();
            if (isRed ? (rDist[node] != -1) : (bDist[node] != -1)){
                continue;
            }
            if (isRed){
                rDist[node] = d;
                for (int v : bEdgeList[node]){
                    if (bDist[v] == -1){
                        q.push(d + 1);
                        q.push(v);
                        stq.push(false);
                    }
                }
            } else {
                bDist[node] = d;
                for (int v : rEdgeList[node]){
                    if (rDist[v] == -1){
                        q.push(d + 1);
                        q.push(v);
                        stq.push(true);
                    }
                }
            }
        }
        vector<int> res(n);
        for (int i = 0; i < n; ++i){
            if (rDist[i] == -1){
                res[i] = bDist[i];
            } else if (bDist[i] == -1){
                res[i] = rDist[i];
            } else {
                res[i] = min(rDist[i], bDist[i]);
            }
        }
        return res;
    }
};
